### DISCUSSION TITLE
iw3 torch_compile patch for Nvidia graphic

### DISCUSSION THREAD

**Discussion by laiyangli001** [nagadomi/nunif#488] [Ideas]
[iw3 torch_compile patch for Nvidia.zip](https://github.com/user-attachments/files/22525462/iw3.torch_compile.patch.for.Nvidia.zip)
Hi, this patch is made by my member of AI 2D to 3D group, can you check if can add into iw3?
<img width="1585" height="311" alt="QQ_1758906667855" src="https://github.com/user-attachments/assets/a5d76683-b635-4441-a640-445bda890a78" />
<img width="1404" height="311" alt="40" src="https://github.com/user-attachments/assets/97f3e52a-7b41-4f83-b3bd-4e9eb0cab175" />

<img width="1233" height="1051" alt="QQ_1758906696367" src="https://github.com/user-attachments/assets/f47f1db6-1e64-4554-a6d1-e0711141b27f" />
Without torch.colpile for nvidia, the fps is 30fps,
With torch.compile for nvidia, the fps is 40fps,
increase 10fps!

**Comment by nagadomi** [comment]
hi,
I won't be able to check for about a month since I'll be in an environment without a computer, but I have already confirmed that torch.compile works on Windows.
The documentation is missing, but no changes to the source code are necessary.

**Comment by nagadomi** [comment]
The following is my assumed procedure based on my memory.

1. Use the latest windows_package from https://github.com/nagadomi/nunif/tree/master/windows_package (nunif-window.zip has not been updated for a long time and is out of date)
2. install pyhton-dev from https://github.com/nagadomi/nunif/tree/master/windows_package/torch_compile. Also apply enable long path
4. install visual-studio 2019 or 2022(C++ desktop env). Also install the English language pack (required). These are set automatically with setenv.bat https://github.com/nagadomi/nunif/blob/master/windows_package/setenv.bat
5. reboot

Now  torch.compile on the CPU should work.
torch.compile with CUDA used to require toriton-windows, but now it seems to work with the official triton.
https://github.com/woct0rdho/triton-windows

Let me know if there are anything missing or if there are any major differences from your method.

**Reply by francdn** [reply]
install_python_dev.bat did not work well for me. I had python already installed but it would output the error "Python not installed". I had to bypass that check on the bat file to make it work. A part from that it would break because the tmp folder inside the windows_package folder did not exist so I have to create the tmp folder manually to make the bat file work.

**Reply by nagadomi** [reply]
It expects the same folder structure.
```
torch_compile/install_python_dev.bat
setenv.bat
update.bat
....
```
It is also expected that the initial setup has been completed by running install.bat or update.bat, making normal iw3 workable.
If you are already using the nunif-windows package, copy the torch_compile folder there.

**Reply by francdn** [reply]
I was able to complete the installation with no problems but torch compile would not enable on the UI when clicked. In any case, I just tried with the patch from laiyangli001 and that made it work.

**Reply by lc700x** [reply]
> install_python_dev.bat did not work well for me. I had python already installed but it would output the error "Python not installed". I had to bypass that check on the bat file to make it work. A part from that it would break because the tmp folder inside the windows_package folder did not exist so I have to create the tmp folder manually to make the bat file work.

This `torch_compile/install_python_dev.bat` only works on CPU. Actually for NVIDIA GPU, it just needs to additionally add this to installation for PyTorch 2.7 to make it work.
```bash
pip install triton-windows<3.4
```

**Reply by laiyangli001** [reply]
yes,I test with the patch, it works
<img width="1228" height="1054" alt="QQ_1758905269812" src="https://github.com/user-attachments/assets/1530b56c-9678-4978-97b5-f5042a936792" />
<img width="1622" height="310" alt="QQ_1758905239284" src="https://github.com/user-attachments/assets/b5dafabb-762a-4a98-89ec-1dfb6b8692ac" />
Does iw3 include TensorRT？Although the fps reaches 50fps,but the desktop screen don't seem very smooth(This condition is for 4K screen capture, 4K streamer output.). Could you optimize them using GPT.

**Reply by laiyangli001** [reply]
Check this project,It is programed and optimized with Chat GPT
https://github.com/lc700x/desktop2stereo

**Reply by laiyangli001** [reply]
<img width="1878" height="540" alt="60" src="https://github.com/user-attachments/assets/fc539bbc-76fc-4e80-8af1-60b4f369d838" />
<img width="1672" height="460" alt="518" src="https://github.com/user-attachments/assets/c82cc2f7-2a90-4787-99ed-b146b1a930e1" />

=====================================================================
 
        self.cbo_stream_fps = EditableComboBox(self.grp_network, choices=["60", "30", "24", "15", "8"],

        self.lbl_stream_height = wx.StaticText(self.grp_network, label=T("Streaming Resolution"))
        self.cbo_stream_height = EditableComboBox(self.grp_network, choices=["2160", "1440", "1080", "720"],
   
      =====================================================================
              self.cbo_resolution = EditableComboBox(self.grp_stereo,
                                               choices=["Default", "196", "294", "336", "392", "448", "490", "518"],
        ====================================================================

can you add these options into gui.py of desktop please?

**Reply by francdn** [reply]
I installed  visual studio 2022(C++ desktop env) and the patch works for me.

**Reply by nagadomi** [reply]
>pip install triton-windows<3.4

Yes, when I tried it before, installing triton-windows was required, so I've included the link above.
However, the triton-windows GitHub page now　states the following:
>This is officially supported by Triton

If this is correct, I think it should work without installing triton-windows by setting up the CUDA compiler or updating torch, but I haven't tried it myself so I don't know.

**Reply by John9520** [reply]
No problem for install torch compile but when I launch the the 2d to 3d conversion with iw3 desktop I got a peak to the down at the level  to GPU a few seconds per minute and the screen freeze for 3-4 seconds before come back.

**Reply by nagadomi** [reply]
The first time takes a long compile time. after the second time, the cache is used, but it still freezes a little. If settings are changed or the video resolution is different, it may be recompiled.

**Reply by nagadomi** [reply]
Currently, iw3.desktop runs garbage collection about once every 30 seconds, and it also resets the Dynamo state. If you see frequent freezes, this might be the reason.
This is only done as a precaution and should not really be necessary.

**Comment by DepthAuteur** [comment]
Hello folks,
Please allow me a short question. Will this also speed up a non realtime 2D to 3D-conversion with the new inpainting method or does this only affect real time streaming?
Thanks and best wishes... TeeJay-NLD

**Reply by nagadomi** [reply]
Yes, normal video conversion also becomes faster.
In a benchmark of the inpainting model alone, the FPS is about 2x higher with torch.compile on Linux.
However, I only use torch.compile during training on Linux, so I'm not very confident it works correctly in other cases but it probably works.
You should be a bit careful in case something in the results is broken.

**Reply by diverswan9** [reply]
@TeeJay-NLD

I tried it (the zip at top). For normal conversions I did get a slight increase. It may vary depending on your PC specs and which size of depth model. Not a big leap in speed but every little bit helps.

**Comment by AscendancyDotA** [comment]
i already updated the embedded python to 3.12 how can i enable torch compile?

**Reply by diverswan9** [reply]
download the zip...
https://github.com/user-attachments/files/22525462/iw3.torch_compile.patch.for.Nvidia.zip
run **Enable Torch_compile.bat**
When you 1st click on **torch.compile** in the GUI it takes a while and then when you run a conversion with it **'on'** it takes a while for it to get going but eventually it gets up to speed. It's not a big difference but for me it's a little faster. It's more noticeable with iw3-desktop.

**Reply by AscendancyDotA** [reply]
it uses python 3.10 in the files

**Reply by diverswan9** [reply]
ah... I see your point. :)

**Reply by nagadomi** [reply]
>I haven't built the dev package for Embeddable Python 3.12 yet, so it doesn't support torch.compile.
>https://github.com/nagadomi/nunif/pull/524#issuecomment-3508536341

[Windows embeddable Python doesn't provide header files or lib](https://github.com/python/cpython/issues/82405), so I need to build them myself.
Since I built the previous python310-dev.zip myself, I should be able to build them the same way, but I haven't done it yet.

**Reply by RudiRastelli** [reply]
how to enable torch compile on ubuntu ?

**Reply by nagadomi** [reply]
It is enabled by default on Linux.
If it cannot be checked, there is a bug.

**Reply by RudiRastelli** [reply]
Yep... it can not be checked in gui

**Reply by nagadomi** [reply]
Device = All CUDA is not supported. Try specifying CPU or single GPU.
In all cases, is it impossible to check the checkbox?

**Reply by RudiRastelli** [reply]
**CPU** - Terminal Output:
```
tmp/tmpt_cmeaqc/main.c:5:10: fatal error: Python.h: file or folder not found
    5 | #include <Python.h>
      |          ^~~~~~~~~~
compilation terminated.
```

**0:NVIDIA GeForce RTX 3060** - Terminal output:
```
W1130 11:58:30.207000 4362 .venv/lib/python3.12/site-packages/torch/_inductor/utils.py:1250] [0/1] Not enough SMs to use max_autotune_gemm mode
/tmp/tmpywn807aa/main.c:5:10: fatal error: Python.h: file or folder not found
    5 | #include <Python.h>
      |          ^~~~~~~~~~
compilation terminated.
```

**Reply by nagadomi** [reply]
Ah, I see. If Python is the system default, install the dev package with `sudo apt-get install python3-dev`.

**Reply by RudiRastelli** [reply]
thx... this works just fine on iw3 and iw3-desktop as well

**Comment by diverswan9** [comment]
https://github.com/nagadomi/nunif/pull/524#issuecomment-3658626940
I ran **update.bat,** ran **install_python_dev.bat** from torch_compile folder, was getting an error so I ran **Enable Torch_compile.bat** from Main folder. From a quick test it worked but the increase was pretty minimal (fps). I'm not sure if this is the correct procedure to get this working for python 3.12...?

**Reply by nagadomi** [reply]
`Enable Torch_compile.bat` was not prepared by me, and it will overwrite the dev files for the old Python 3.10. Please do not run it.
`torch_compile/install_python_dev.bat` is not updated by `update.bat`. It is updated by `update-installer.bat`.

However, if you were able to check torch.compile, then it is working. Performance is not my responsibility, but the regular iw3-gui includes the long compile time in its FPS calculation, so the displayed FPS is underestimated and gradually increases over time. This issue does not occur in iw3-desktop.
